maitake/src/wait/queue.rs

```diff
@@ -0,0 +1,790 @@
+    Closed,
+}
+
+// === impl WaitQueue ===
+
+impl WaitQueue {
+    /// Returns a new `WaitQueue`.
+    #[must_use]
+    #[cfg(not(loom))]
+    pub const fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(0)),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Returns a new `WaitQueue`.
+    #[must_use]
+    #[cfg(loom)]
+    pub fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(0)),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Wake the next task in the queue.
+    ///
+    /// If the queue is empty, a wakeup is stored in the `WaitQueue`, and the
+    /// next call to [`Wait`] will complete immediately.
+    #[inline]
+    pub fn wake(&self) {
+        // snapshot the queue's current state.
+        let mut state = self.load();
+
+        // check if any tasks are currently waiting on this queue. if there are
+        // no waiting tasks, store the wakeup to be consumed by the next call to
+        // `wait`.
+        loop {
+            match state.get(QueueState::STATE) {
+                // if the queue is closed, bail.
+                State::Closed => return,
+                // if there are waiting tasks, break out of the loop and wake one.
+                State::Waiting => break,
+                _ => {}
+            }
+
+            let next = state.with_state(State::Woken);
+            // advance the state to `Woken`, and return (if we did so
+            // successfully)
+            match self.compare_exchange(state, next) {
+                Ok(_) => return,
+                Err(actual) => state = actual,
+            }
+        }
+
+        // okay, there are tasks waiting on the queue; we must acquire the lock
+        // on the linked list and wake the next task from the queue.
+        let mut queue = self.queue.lock();
+        test_trace!("wake: -> locked");
+
+        // the queue's state may have changed while we were waiting to acquire
+        // the lock, so we need to acquire a new snapshot.
+        state = self.load();
+
+        if let Some(waker) = self.wake_locked(&mut *queue, state) {
+            drop(queue);
+            waker.wake();
+        }
+    }
+
+    /// Wake *all* tasks currently in the queue.
+    pub fn wake_all(&self) {
+        let mut queue = self.queue.lock();
+        let state = self.load();
+
+        match state.get(QueueState::STATE) {
+            // if the queue is closed, bail.
+            State::Closed => return,
+
+            // if there are no waiters in the queue, increment the number of
+            // `wake_all` calls and return.
+            State::Woken | State::Empty => {
+                self.state.fetch_add(QueueState::ONE_WAKE_ALL, SeqCst);
+                return;
+            }
+            State::Waiting => {}
+        }
+
+        // okay, we actually have to wake some stuff.
+
+        // TODO(eliza): wake outside the lock using an array, a la
```

Comment:
Can this todo be closed? I don't really track...

Reply:
this is a thing i still want to do in a follow-up PR.